### ISSUE TITLE
It is unclear how to convert "source" and "target" parameters to URIs

### ISSUE THREAD

**Issue by melvincarvalho** [aaronpk/webmention#9]
In section 5 of the protocol

> Barnaby's server sends a webmention to Aaron's post's webmention endpoint with

    source set to Barnaby's post's permalink
    target set to Aaron's post's permalink.

Source and target are form encoded parameter.  While this makes sense in that context it suffers from a couple of weaknesses.

- Similar semantics may be wished to be sent with another mime type (e.g. using those developed by the social web WG, such as JSON based or other w3c rec)
- In other scenarios "source" and "target" as simple strings my be subject to name clashes and collisions

In order to make this more scalable it is advantageous to be able to systematically convert those parameters into URIs.  This information could be gained either in a generic way that applies to all form encoded variables (tho none exist today), or described in the spec.

It is undesirable for software to do this ad hoc, as different decisions might be made by different code bases leading to interoperability issues.

Suggestions for this:

- use the pingback vocab ( http://www.w3.org/wiki/Pingback#Semantic_Pingback )
- use a URN
- create a new vocab, e.g. at w3.org/ns or w3id.org
- other

If there's no preference here, I'd suggest using the pingback vocab as it was one of the cited motivations for webmention.  

A few words in the text of the spec could make clear how software providers could use webmention in the form context and also with linked data based systems.

**Comment by bblfish** [comment]
I suppose by *the protocol* you mean [webmention](http://indiewebcamp.com/webmention). It's always good to add links, as people can come to a page from any context. 

As it happens, that is the problem with urlencoded key/value pair forms that lack a global namespace: since people/agents can come to a resource from any other server - we are in a p2p global information space after all - there has to be a way for the client to be able to be clear when he is posting something, what the meaning of what is being posted is, in such a way that there cannot be a disagreement with the server. Otherwise the danger is that someone would set up a **joining the army form**, and use exactly the same parameter names as webmention, leading to a lot of people joining the army against their will.
This was known as [taking the king's shilling](http://www.royalscotsgrenadiers.com/kingsshilling.html).

This is even more important when we are creating software based User Interfaces that follow relations around, as they won't be able to take the context, aesthetics and meaning of the page into account before sending a page. These agents therefore need to know before posting what the meaning of what the content they will `POST` is. 

1. One way to do this is to make sure the mime type has a well understood interpretation.

2. Another way is to extend urlencoding by turning attributes into URLs as proposed by @melvincarvalho .  This could function in that it could be argued that the meaning of urlencoded forms until now was always a client/server relation and that the parameters therefore were always interpreted relative to the forms base url.

3.  The other way to do this, is for the resource to specify in advance its interpretation of the meaning of what will be sent, given that what is sent does not have a well understood meaning.

This suggests that what is called an *endpoint* in the webmention spec, and would be thought of as an LDP Container in LDP, define what type of content can be sent to it, and how that is interpreted into a language that has well defined semantics (eg. RDF).

This may not be that difficult to do. We could define a new `Link:` relation say `urlencoded` that would point to a transformer from urlencoding to rdf. Say something like this (see [Web Linking RFC](https://tools.ietf.org/html/rfc5988))

```
200 Ok
Link: <http://w3c.org/social/WebMention>; rel=urlencoded
```

The document at  `<http://w3c.org/social/WebMention>` would have both an HTML representation and a machine readable representation. The machine readable representation would just explain how webmention works, the header, and the machine transformation. It may just need to be an attribute value pair file, that for each attribute in the form assigns the URL of the relation that it stands for.  One would then still need to work out how to express in a machine readable form what the subject of the form is.

What one really wants is a simple language like this

```
PREFIX ping: <http://purl.org/net/pingback/>
CONSTRUCT { 
  [] ping:source ?source;
     ping:target ?target . 
} FROM ?source ?target
```

Where `?source` and `?target` are the attribute names of the form.  This is not that complicated and would allow us to de-siloeify all forms on the web.

This would allow the IndieWeb folk to increase the security of their protocol while retaining their principle of remaining accessible, and it would allow this to be integrated generically into the SoLiD platform, so as to reduce configuration mistakes, and make it easier to automatically create such resources. This would require from the LDPnext side to work out how one can increase the mime type to such a urlencoded form.

**Comment by bblfish** [comment]
there is still a problem with this proposal in that it is not clear what the subject of the attribute/values is.

**Comment by gobengo** [comment]
@melvincarvalho Would it resolve your concern (and this issue) to add something like this (exact verbage left to editor) to the specification?

"Webmention-receiving services that store or republish the 'source' and 'target' properties as URIs SHOULD do so using [http://purl.org/net/pingback/source](http://dssn.org/pingback/ns/namespace.html#idp183472) and [http://purl.org/net/pingback/target](http://dssn.org/pingback/ns/namespace.html#idp267488)."

**Comment by bblfish** [comment]
@gobengo 

- it's unlikely that this will be accepted by the IndieWeb folks, as they will argue that it is ugly
- it partially answers the problem here but not fully as per my previous remark that it is not clear what the subject of the request is
- it won't scale to more cases ( see issue #11 ), meaning that there would be even less reason for it to be adopted by the IndieWeb folks, nor for it to make that much sense to the w3c.

**Comment by rhiaro** [comment]
I was gonna suggest similar to @gobengo, plus the `webmention` rel value could be mapped to `pingback:to`.

@bblfish The spec says upon receipt of a webmention a URL must be returned, so the subject of `source` and `target` would presumably be that URL - the 'webmention object' (a `pingback:Request`?). (will send a longer reply to #11 later).

**Comment by melvincarvalho** [comment]
@gobengo assuming that's the ns used, that would work yes.  However, there's talk of purl.org being revamped.  So it would be good to get consensus on where to put it.  But in principle, yes, that sounds like improved text.

@bblfish agree with all your points.  But the subject issue was not raised here, let's follow that up in #11 and perhaps try and resolve just "source" and "target" in this thread.  Re: "indieweb folks" accepting it, Im an indieweb folk and so is amy, so that's a good start.

@rhiaro the subject of the webmention is something like 

    <#subject>  <:source>  <URL>

Now if it's not mentioned we can assume it's an anonymous or blank node.  That's ok as linked data permits blank nodes, tho many consider it an anti pattern.

**Comment by gobengo** [comment]
@rhiaro 
> the subject of source and target would presumably be that URL - the 'webmention object' (a pingback:Request?). (will send a longer reply to #11 later).

It seems to be that [`pingback:Item`](http://dssn.org/pingback/ns/namespace.html#idp183472) is more fitting than `pingback:Request`, which is linked to as http://purl.org/net/pingback/Request, but 404s

@melvincarvalho 
> However, there's talk of purl.org being revamped. So it would be good to get consensus on where to put it.

Cool. I suppose I'm implicitly proposing trying to make use of the existing pingback vocab instead of introducing a new one (until/unless we see a great reason the pingback vocab won't be sufficient). So hopefully nothing new will be created that has to be put somewhere new.

To summarize so far it seems like
* This is a reasonable issue to try consider
* It can be addressed by adding one sentence, probably just pointing to an existing vocabulary. No new MUSTs required by webmention.

**Comment by rhiaro** [comment]
@gobengo For when purl.org is down, the actual SP vocab is here: http://dssn.org/pingback/ns/namespace.html. Probably not the best place to be debating the semantics of the terms, but `pingback:Item` has seeAlso to `pingback:Container` (which I *believe* is a type for the endpoint itself?) so that doesn't feel right to me, although none of them have descriptions so it's not clear what's intended.

**Comment by gobengo** [comment]
> For when purl.org is down, the actual SP vocab is here

Right by the namespace URI is still `http://purl.org/net/pingback/`. I agree that the canonical document that describes that namespace can/should be somewhere that stays highly available.

> pingback:Item has seeAlso to pingback:Container

Oh I assumed 'seeAlso' was just 'this is relevant' not 'this is equivalent'.

I think we're essentially agreeing so happy to let this rest.

**Comment by azaroth42** [comment]
Can someone clarify the distinction with #10 please? This issue is /only/ about republishing once received?

**Comment by bblfish** [comment]
This is a particular answer to the problem discussed in #10

**Comment by aaronpk** [comment]
@jasnell what was the mechanism/language you used to use a default namespace (base URI) for property names in Activity Streams? I'd like to add a similar paragraph to Webmention to resolve this issue.

**Comment by aaronpk** [comment]
Will add this to the appendix:

```
If your implementation wants these to be a URI, prefix them with http://w3c.org/ns/webmention#
```

**Comment by sandhawke** [comment]
Minor correction: http://www.w3.org/ns/webmention#

**Comment by aaronpk** [comment]
oops thanks!

**Comment by aaronpk** [comment]
Added! http://webmention.net/draft/#uris-for-form-encoded-properties

**Comment by aaronpk** [comment]
@melvincarvalho We discussed this issue during the March F2F meeting. I've updated the spec to address the issue, giving guidance on how to turn the parameters into URIs. Please take a look at the [revised text here](http://webmention.net/draft/#uris-for-form-encoded-properties) and comment here whether this is acceptable. If not, you may reopen the issue and provide more information about what change is requested and why.

**Comment by melvincarvalho** [comment]
@aaronpk thanks, that looks like an improvement, and it would help resolve interop issues.  

First comment is that ( http://www.w3.org/ns/webmention# ) gives a 404 right now.  

What's the consensus regarding reusing the pingback ontology, as I think it already contains all the terms used in webmention?

I think @csarven has done some work here on the solid notifications spec and implementation, and may wish to weigh in.

**Comment by jasnell** [comment]
@aaronpk ... my apologies, I just saw the mention on this and the question about the default namespace.  I'm not 100% certain what you're asking for tho. Can you clarify?

**Comment by aaronpk** [comment]
@jasnell thanks, we discussed during the f2f and came up with the language I added to the appendix.

@melvincarvalho I can't find a reference to namespace used for the Pingback terms. Can you point me to it?

@sandhawke Is there something we need to put at http://www.w3.org/ns/webmention for this? How would we go about that?

**Comment by rhiaro** [comment]
@aaronpk @sandhawke I'm happy to take charge of putting the LD-friendly representation of terms up at www.w3.org/ns/webmention

@melvincarvalho SP has `source` and `target` but also a bunch of other constraints that might not apply to webmention. For people interested in the LD side of this we should probably have a separate discussion sometime. We can always include equivalence to related terms in the wm namespace if appropriate.
